core/trino-main/src/main/java/io/trino/cost/AggregationStatsRule.java

```diff
@@ -50,36 +51,47 @@ public Pattern<AggregationNode> getPattern()
     @Override
     protected Optional<PlanNodeStatsEstimate> doCalculate(AggregationNode node, StatsProvider statsProvider, Lookup lookup, Session session, TypeProvider types, TableStatsProvider tableStatsProvider)
     {
-        if (node.getGroupingSetCount() != 1) {
+        if (node.getGroupingSetCount() != 1 || node.getStep() == INTERMEDIATE) {
             return Optional.empty();
         }
 
-        if (node.getStep() != SINGLE && node.getStep() != FINAL) {
-            return Optional.empty();
+        PlanNodeStatsEstimate estimate;
+
+        if (node.getStep() == PARTIAL) {
+            estimate = getPartialEstimate(statsProvider.getStats(node.getSource()),
+                    node.getGroupingKeys(),
+                    node.getAggregations());
+        }
+        else {
+            estimate = getEstimate(
+                    statsProvider.getStats(node.getSource()),
+                    node.getGroupingKeys(),
+                    node.getAggregations());
+        }
+        return Optional.of(estimate);
+    }
+
+    public static PlanNodeStatsEstimate getPartialEstimate(PlanNodeStatsEstimate sourceStats, Collection<Symbol> groupBySymbols, Map<Symbol, Aggregation> aggregations)
+    {
+        PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.builder();
+        // Pessimistic assumption of no reduction from PARTIAL aggregation, forwarding of the source statistics.
+        result.setOutputRowCount(sourceStats.getOutputRowCount());
+        result.addSymbolStatistics(getGroupBySymbolsStatistics(sourceStats, groupBySymbols));
+        for (Map.Entry<Symbol, Aggregation> aggregationEntry : aggregations.entrySet()) {
+            result.addSymbolStatistics(aggregationEntry.getKey(), estimateAggregationStats(aggregationEntry.getValue(), sourceStats));
         }
 
-        return Optional.of(groupBy(
-                statsProvider.getStats(node.getSource()),
-                node.getGroupingKeys(),
-                node.getAggregations()));
+        return result.build();
     }
 
-    public static PlanNodeStatsEstimate groupBy(PlanNodeStatsEstimate sourceStats, Collection<Symbol> groupBySymbols, Map<Symbol, Aggregation> aggregations)
+    public static PlanNodeStatsEstimate getEstimate(PlanNodeStatsEstimate sourceStats, Collection<Symbol> groupBySymbols, Map<Symbol, Aggregation> aggregations)
```

Comment:
This is an even less descriptive name than before. We can use the existing name or estimateGroupBy.
You can note in a java doc comment that this method is used to estimate FINAL or SINGLE step aggregations.